Rewrite `attr()` to match the newest version of the spec

bramus opened the pull request:
_(👋  Hi, Chrome DevRel here)_

### Description

The `attr()` function underwent an upgrade: https://drafts.csswg.org/css-values-5/#attr-notation
This PR updates the `attr()` page with that info.

### Motivation

This feature is shipping in the upcoming Chrome 133: https://chromestatus.com/feature/4680129030651904

### Additional details

- This is my first major contribution to MDN, so most likely I got a bunch of things wrong. I modeled the page structure after [`calc-size()`](https://developer.mozilla.org/en-US/docs/Web/CSS/calc-size)
- There are still two things marked as `TODO:` in the file, for which I could use some guidance.

### Related issues and pull requests

There is still an issue with the data table, for which I have file https://github.com/mdn/browser-compat-data/issues/25618

bramus committed:
Rewrite attr() to match the newest version of the spec

chrisdavidmills committed:
Merge branch 'main' into advanced-attr

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -22,126 +22,44 @@ attr(data-count);
+- `<attr-type>`
+  - : Specifies how the attribute value is parsed into a CSS value.
+- `<declaration-value>`
+  - : The value to be used if the associated attribute is missing or contains an invalid value. If not set, CSS will default to an empty string when no `<attr-type>` is set or the guaranteed-invalid value.
```
```suggestion
  - : The value to be used if the specified attribute is missing or contains an invalid value. If not set, CSS will default to an empty string when no `<attr-type>` is set or the guaranteed-invalid value.
```

chrisdavidmills replied:
My suggestion only covers the first sentence.

I'd remove the second sentence and instead include this information in the "Return value" section, as it is really a description of what the return value ends up as, if the fallback is or isn't included and the `<attr-type>` is or isn't included, right? But I also think it needs work. For example, it is no use saying "guaranteed-invalid value", as most readers won't know what that means. Either add more description to this page, or add an MDN glossary entry for it and link to it?

bramus replied:
Added a glossary entry.

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -22,126 +22,44 @@ attr(data-count);
-      - : The attribute value is parsed as a CSS {{CSSxRef("&lt;number&gt;")}}, that is without the unit (e.g. `12.5`), and interpreted as a {{CSSxRef("&lt;percentage&gt;")}}. If it is not valid, that is not a number or out of the range accepted by the CSS property, the default value is used.
-        If the given value is used as a length, `attr()` computes it to an absolute length.
-        Leading and trailing spaces are stripped.
+### Return Value
```
```suggestion
### Return value
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -22,126 +22,44 @@ attr(data-count);
+### Return Value
 
-        Default value: `0%`, or, if `0%` is not a valid value for the property, the property's minimum value.
+By default attributes will be parsed into a CSS string but you can change that with the optional `<attr-type>`. You can set `<attr-type>` to one of these three:
```
```suggestion
By default attributes will be parsed into a CSS string but you can change that with the optional `<attr-type>`. You can set `<attr-type>` to one of the following:
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -22,126 +22,44 @@ attr(data-count);
 
-- `<fallback>`
-  - : The value to be used if the associated attribute is missing or contains an invalid value. If not set, CSS will use the default value defined for each `<type-or-unit>`.
+- The `string` keyword which parses the value into a CSS string. This has the same outcome as omitting the `<attr-type>`.
```
```suggestion
- The `string` keyword, which parses the value into a CSS string. This has the same outcome as omitting the `<attr-type>`.
```

chrisdavidmills replied:
But, as I said above, I'd include the detailed information on `<attr-type>` values in a "Description" section. This section should just list the possible Return value types, and say briefly how they are achieved.

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -22,126 +22,44 @@ attr(data-count);
-- `<fallback>`
-  - : The value to be used if the associated attribute is missing or contains an invalid value. If not set, CSS will use the default value defined for each `<type-or-unit>`.
+- The `string` keyword which parses the value into a CSS string. This has the same outcome as omitting the `<attr-type>`.
+- The `type()` function which needs a `<syntax>` to know what [data type](/en-US/docs/Web/CSS/CSS_Types) to parse the value into. Possible `<syntax>`es include {{CSSxRef("&lt;angle&gt;")}}, {{CSSxRef("&lt;color&gt;")}}, {{CSSxRef("&lt;custom-ident&gt;")}}, {{CSSxRef("&lt;image&gt;")}}, {{CSSxRef("&lt;integer&gt;")}}, {{CSSxRef("&lt;length&gt;")}}, {{CSSxRef("&lt;length-percentage&gt;")}}, {{CSSxR ...
```
```suggestion
- The `type()` function, which takes a `<syntax>` as its argument that specifies what [data type](/en-US/docs/Web/CSS/CSS_Types) to parse the value into. The `<syntax>` can be {{CSSxRef("&lt;angle&gt;")}}, {{CSSxRef("&lt;color&gt;")}}, {{CSSxRef("&lt;custom-ident&gt;")}}, {{CSSxRef("&lt;image&gt;")}}, {{CSSxRef("&lt;integer&gt;")}}, {{CSSxRef("&lt;length&gt;")}}, {{CSSxRef("&lt;length-percentage&gt;")}}, {{CSSxRef("&lt;number&gt;")}}, {{CSSxRef("&lt;percentage&gt;")}}, {{CSSxRef("&lt;resolution&gt;")}}, {{CSSxRef("&lt;string&gt;")}}, {{CSSxRef("&lt;time&gt;")}}, and {{CSSxRef("&lt;transform-function&gt;")}}, or a combination thereof. Excluded from this list is {{CSSxRef("&lt;url&gt;")}} for [security reasons](#limitations_and_security)
```

chrisdavidmills replied:
This description could really do with some example snippets to give the reader an idea of what is acceptable, especially for combinations of multiple syntaxes. Which combinations are valid?

bramus replied:
Added examples

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -22,126 +22,44 @@ attr(data-count);
-  - : The value to be used if the associated attribute is missing or contains an invalid value. If not set, CSS will use the default value defined for each `<type-or-unit>`.
+- The `string` keyword which parses the value into a CSS string. This has the same outcome as omitting the `<attr-type>`.
+- The `type()` function which needs a `<syntax>` to know what [data type](/en-US/docs/Web/CSS/CSS_Types) to parse the value into. Possible `<syntax>`es include {{CSSxRef("&lt;angle&gt;")}}, {{CSSxRef("&lt;color&gt;")}}, {{CSSxRef("&lt;custom-ident&gt;")}}, {{CSSxRef("&lt;image&gt;")}}, {{CSSxRef("&lt;integer&gt;")}}, {{CSSxRef("&lt;length&gt;")}}, {{CSSxRef("&lt;length-percentage&gt;")}}, {{CSSxR ...
+- A `<attr-unit>` which is any identifier that matches a [CSS distance unit](/docs/Web/CSS/CSS_Values_and_Units#distance_units), such as `px`, or the `%` character.
```
```suggestion
- An `<attr-unit>`, which is an identifier that matches a [CSS distance unit](/docs/Web/CSS/CSS_Values_and_Units#distance_units) such as `px` or `rem`, or the `%` character.
```

chrisdavidmills replied:
It can also be other unit types, such as resolution, time, angle, etc., can't it?

If the Chrome 133 release only supports a subset of the specified allowed units, then the BCD should specify `"partial_implementation": true`, and include a note to specify what is currently supported.

bramus replied:
> It can also be other unit types, such as resolution, time, angle, etc., can't it?

Correct. Added more examples.

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -200,12 +118,13 @@ In this example, we set the CSS value of {{CSSXRef("background-color")}} to the
+// TODO: This live Example needs to be adjusted
 {{EmbedLiveSample("color_value", "100%", 50)}}
 
 ### using fallback
```
```suggestion
### Using a fallback value
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -200,12 +118,13 @@ In this example, we set the CSS value of {{CSSXRef("background-color")}} to the
 
 #### Result
 
+// TODO: This live Example needs to be adjusted
```
What is the issue with this live example? I'm afraid I can't look at it because the live examples seem to be b0rked in the live preview pages, but I'm assuming you've tested it locally.

bramus replied:
It uses an older syntax. I am not sure where I can find this live example so that I can edit it.

bramus replied:
Perfect, thanks!

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -234,6 +153,142 @@ p::after {
 
+### `<custom-ident>`
+
+In this example the values for the {{cssxref("view-transition-name")}} property are dereived from the `id` attribute of the element. The attribute gets parsed into a {{CSSxRef("&lt;custom-ident&gt;")}} which is what {{cssxref("view-transition-name")}} needs.
```
```suggestion
In this example, the values for the {{cssxref("view-transition-name")}} property are derived from the `id` attribute of the element. The attribute gets parsed into a {{CSSxRef("&lt;custom-ident&gt;")}}, which is what {{cssxref("view-transition-name")}} needs.
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
````diff
@@ -234,6 +153,142 @@ p::after {
+  view-transition-name: attr(id type(<custom-ident>), none);
+  view-transition-class: card;
+}
+```
````
It would be lovely if we could see this expended into a live example.

bramus replied:
Added. Also adding the rotation example

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
````diff
@@ -234,6 +153,142 @@ p::after {
+}
+```
+
+### dimension unit
````
```suggestion
### Using dimension units
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -234,6 +153,142 @@ p::after {
 
 {{EmbedLiveSample("using_fallback", "100%", 90)}}
 
+### `<custom-ident>`
```
```suggestion
### Parsing `attr()` values as `<custom-ident>`s
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -234,6 +153,142 @@ p::after {
+
+### dimension unit
+
+In this example the `data-size` attribute gets parsed into a pixel value.
```
Again, it might be nice to see a slightly more unusual value used, such as `rem` or `vw`. Also, it would be nice to to expand this a bit into a live example. The example in the spec is kinda nice, although it would be better without the custom element names.

bramus replied:
Updated the example.

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -234,6 +153,142 @@ p::after {
+
+## TODO: Title for this section
+
+### Limitations and Security
```
These H3s should be subsections inside the "Description" section that I mentioned earlier.

chrisdavidmills replied:
```suggestion
### Limitations and security
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -234,6 +153,142 @@ p::after {
+
+### Limitations and Security
+
+The `attr()` function can reference attributes that were never intended by the page to be used for styling, and might contain sensitive information (for example, a security token used by scripts on the page). In general this is fine, but it can become a security threat when used in URLs. Therefore you can't use `attr()` to dynamically construct URLs.
```
```suggestion
The `attr()` function can reference attributes that were never intended by the page author to be used for styling, and might contain sensitive information (for example, a security token used by scripts on the page). In general, this is fine, but it can become a security threat when used in URLs. You therefore can't use `attr()` to dynamically construct URLs.
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
````diff
@@ -234,6 +153,142 @@ p::after {
+</style>
+```
+
+Values that use get marked as _"`attr()`-tained"_. Using an `attr()`-tainted value as or in a `<url>` makes a declaration become ["invalid at computed value time" or IACVT for short](https://brm.us/iacvt)
````
```suggestion
`attr()` return values that include URLs get marked as _"`attr()`-tained"_. Using an `attr()`-tainted value as or within a `<url>` makes a declaration become ["invalid at computed value time" or IACVT for short](https://brm.us/iacvt)
```

chrisdavidmills replied:
I'm not sure if this is correct, but something was missing from your original sentence.

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -234,6 +153,142 @@ p::after {
+
+### Backwards compatibility
+
+Generally speaking `attr()` is pretty much backwards compatible because the old way of using it–the one without specifying an `<attr-type>`–behaves the same as before. Having `attr(data-attr)` in your code is the same as writing `attr(data-attr type(<string>))` or the simpler `attr(data-attr string))`.
```
```suggestion
Generally speaking, the modern `attr()` syntax is backwards-compatible because the old way of using it — without specifying an `<attr-type>` — behaves the same as before. Having `attr(data-attr)` in your code is the same as writing `attr(data-attr type(<string>))` or the simpler `attr(data-attr string))`.
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -234,6 +153,142 @@ p::after {
+
+Generally speaking `attr()` is pretty much backwards compatible because the old way of using it–the one without specifying an `<attr-type>`–behaves the same as before. Having `attr(data-attr)` in your code is the same as writing `attr(data-attr type(<string>))` or the simpler `attr(data-attr string))`.
+
+However, there are two edge cases where advanced `attr()` behaves differently from the old one.
```
```suggestion
However, there are two edge cases where the modern `attr()` syntax behaves differently from the old syntax.
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -234,6 +153,142 @@ p::after {
+
+However, there are two edge cases where advanced `attr()` behaves differently from the old one.
+
+In the following snippet, browsers with no advanced `attr()` support will discard the second declaration because it cannot parse it. Therefore the result in those browsers is `"Hello World"`.
```
```suggestion
In the following snippet, browsers that don't support the modern `attr()` syntax will discard the second declaration because they cannot parse it. The result in those browsers is `"Hello World"`.
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
````diff
@@ -234,6 +153,142 @@ p::after {
+}
+```
+
+In browsers with advanced `attr()` support the output will be … nothing. Those browsers will successfully parse the second declaration but because it is gibberish content for the `content` property, the declaration becomes ["invalid at computed value time" or IACVT for short](https://brm.us/iacvt).
````
```suggestion
In browsers with support for the modern syntax, the output will be … nothing. Those browsers will successfully parse the second declaration but, because it is invalid content for the `content` property, the declaration becomes ["invalid at computed value time" or IACVT for short](https://brm.us/iacvt).
```

chrisdavidmills replied:
I'm intrigued by this example. Sure if the second declaration is invalid, the first one is still valid? Or does IACVT still provide a value but then it doesn't work? Might be worth clarifying this a bit.

bramus replied:
It’s regular cascade: the last declaration wins but since its value is gibberish the entire declaration becomes IACVT. At that point the earlier declaration from the 1st rule was already discarded. This is clarified in the linked to IACVT article.

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
````diff
@@ -234,6 +153,142 @@ p::after {
+}
+```
+
+In browsers with no advanced `attr()` support the text `"foo"` gets displayed. In browsers with `attr()` support there is no output.
````
```suggestion
In browsers without support for the modern syntax, the text `"foo"` gets displayed. In browsers with modern `attr()` support there is no output.
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -234,6 +153,142 @@ p::after {
+
+In browsers with no advanced `attr()` support the text `"foo"` gets displayed. In browsers with `attr()` support there is no output.
+
+That is because `attr()`–similar to custom properties that use the `var()` function–get substituted at [computed value time](https://brm.us/iacvt/#custom-properties). With the new behavior, `--x` first tries to read the `data-attr` attribute from the `#parent` element, which results in an empty string because there is no such attribute on `#parent`. That empty string then gets inherited onto the  ...
```
```suggestion
This is because `attr()` — similar to custom properties that use the `var()` function — get substituted at [computed value time](https://brm.us/iacvt/#custom-properties). With the modern behavior, `--x` first tries to read the `data-attr` attribute from the `#parent` element, which results in an empty string because there is no such attribute on `#parent`. That empty string then gets inherited by the `#child` element, resulting in a `content: ;` declaration being set.
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -234,6 +153,142 @@ p::after {
+
+To prevent this sort of situation, make sure you don't pass inherited `attr()` values onto children, unless you explicitly want to.
+
+### Feature Detection
```
```suggestion
### Feature detection
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -234,6 +153,142 @@ p::after {
+
+### Feature Detection
+
+We can feature detect support for advanced `attr()` using the {{CSSxRef("@supports")}} at-rule by trying it to assign advanced `attr()` to a (non-custom) CSS property.
```
```suggestion
You can feature detect support for modern `attr()` syntax using the {{CSSxRef("@supports")}} at-rule. In the test, try to assign advanced `attr()` to a (non-custom) CSS property.
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
````diff
@@ -234,6 +153,142 @@ p::after {
+
+```css
+@supports (x: attr(x type(*))) {
+  /* Browser has advanced attr() support */
````
```suggestion
  /* Browser has modern attr() support */
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -234,6 +153,142 @@ p::after {
+}
+
+@supports not (x: attr(x type(*))) {
+  /* Browser has no advanced attr() support */
```
```suggestion
  /* Browser does not have modern attr() support */
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
````diff
@@ -234,6 +153,142 @@ p::after {
+
+```js
+if (CSS.supports("x: attr(x type(*))")) {
+  /* Browser has advanced attr() support */
````
```suggestion
  /* Browser has modern attr() support */
```

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -234,6 +153,142 @@ p::after {
+}
+
+if (!CSS.supports("x: attr(x type(*))")) {
+  /* Browser has no advanced attr() support */
```
```suggestion
  /* Browser does not have modern attr() support */
```

chrisdavidmills reviewed (changes requested):
Hi @bramus! This is a really good effort, thanks for updating this page! I've made quite a few comments, but a lot of them are language-style nitpicks; there isn't too much in terms of significant work to do here.

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Use attr(href) as an example

bramus commented on files/en-us/web/css/attr/index.md:
````diff
@@ -22,126 +22,44 @@ attr(data-count);
 
-    - `time`
+```plain
+attr(<attr-name> <attr-type>? , <declaration-value>?)
````
I was following https://developer.mozilla.org/en-US/docs/Web/CSS/calc-size here as a guideline. Will keep only the first line.

bramus committed:
Add an extra unit example

bramus committed:
Rework syntax

bramus committed:
Remove semicolons as per guidelines

bramus committed:
Move description section around

bramus committed:
Clarify the allowed values for `<attr-type>`

bramus commented on files/en-us/web/css/attr/index.md:
```diff
@@ -22,126 +22,44 @@ attr(data-count);
+  - : The name of an attribute on the HTML element referenced in the CSS.
+- `<attr-type>`
+  - : Specifies how the attribute value is parsed into a CSS value.
+- `<declaration-value>`
```
Adjusted.

bramus committed:
Clarify what happens when a value cannot be parsed into the specified <attr-type>

bramus committed:
Define the guaranteed-invalid value term

bramus committed:
Rework return value and examples

bramus committed:
Fix attr()-tainted sentence

bramus committed:
Add more <attr-unit> examples

bramus committed:
Update dimension example

bramus commented:
Thanks for the review, Chris. Merged a bunch of your suggestions and addressed the things I could address. Those that I could not address I have left as unresolved + left a comment.

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -8,140 +8,173 @@ browser-compat: css.types.attr
 
-      - : The attribute value is parsed as a CSS {{CSSxRef("&lt;number&gt;")}}. If it is not valid, that is not a number or out of the range accepted by the CSS property, the default value is used.
-        Leading and trailing spaces are stripped.
+### Return value
```
The structure is much better now, but the "Return value" section now feels a bit incomplete. I think you need to start it with a one-sentence summary of what the return value is in general, followed by a more detailed description of how this value differs depending on the arguments you specify.

You've mostly got this, but you also need to mention the string keyword and the `type()` function.

bramus replied:
Added.

chrisdavidmills reviewed (changes requested):
@bramus much improved! A few more comments for you; getting pretty close now.

bramus committed:
Apply suggestions from code review

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus commented:
Thanks for your input, @chrisdavidmills – Much appreciated! I’ve committed the suggestions you code and am about to go over the remaining conversations now.

chrisdavidmills commented on files/en-us/glossary/guaranteed_invalid_value/index.md:
```diff
@@ -0,0 +1,17 @@
+---
+
+{{GlossarySidebar}}
+
```
This page reads fine as it is, but I think it is missing one thing — what is the Guaranteed-invalid value in general? What is it's purpose, why do we need it, what does it do? This is not clear after reading this page. The page should start with a sentence "The guaranteed-invalid value is..."

bramus replied:
It does start with that on the line below:

```md
In CSS the guaranteed-invalid value is {{CSSXref("initial")}}.
```

chrisdavidmills replied:
Yeah, that says specifically what the value is, but I guess I am still struggling a bit with why this is a specific thing conceptually. Why it is useful to people? Or is it more of a thing to warn people about, not to use this value in `var()` because it is guaranteed invalid?

bramus replied:
I think that exceeds the extent of this PR and would be a good subject for a standalone one.

chrisdavidmills replied:
OK, that's fine. I don't think this is something we need to block this PR on.

chrisdavidmills reviewed (changes requested):
@bramus sorry, I forgot to review the new glossary page. One more new comment added!

bramus committed:
Remove update notice, the live example updates all by itself

bramus committed:
Embed a LiveSample for the view-transition-name example

bramus committed:
Embed a LiveSample for the rotation example

bramus committed:
Add intro sentence to the return value section

bramus committed:
Clarify how to set multiple types

bramus commented on files/en-us/web/css/attr/index.md:
```diff
@@ -8,140 +8,173 @@ browser-compat: css.types.attr
-      - : The attribute value is parsed as a string that is used inside a CSS `url()` function.
-        Relative URL are resolved relatively to the original document, not relatively to the style sheet.
-        Leading and trailing spaces are stripped.
+    - The `type()` function, which takes a `<syntax>` as its argument that specifies what [data type](/en-US/docs/Web/CSS/CSS_Types) to parse the value into. The `<syntax>` can be {{CSSxRef("&lt;angle&gt;")}}, {{CSSxRef("&lt;color&gt;")}}, {{CSSxRef("&lt;custom-ident&gt;")}}, {{CSSxRef("&lt;image&gt;")}}, {{CSSxRef("&lt;integer&gt;")}}, {{CSSxRef("&lt;length&gt;")}}, {{CSSxRef("&lt;length-percent ...
```
Added a small sentence that shows how to combine mutliple types + I split off the example from the code block.

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
````diff
@@ -234,6 +274,174 @@ p::after {
+
+#### Result
+
+```js hidden
````
I would also show the JavaScript in a "JavaScript" section, an explain what it is doing.

bramus replied:
.done

chrisdavidmills commented on files/en-us/web/css/attr/index.md:
```diff
@@ -234,6 +274,174 @@ p::after {
+
+{{EmbedLiveSample("parsing_attr_values_as_custom-idents", "100%", 400)}}
+
+### Using dimension units
```
This example doesn't need so much explanation, as it is really obvious what is going on.

Really nice examples BTW!

chrisdavidmills replied:
I would also consider putting the view transitions one at the bottom of the section, as it is by far the most complex example.

bramus replied:
.done

chrisdavidmills reviewed (approved):
@bramus really nice work. I've added a few more comments, but I'm going to approve this now, as I think it is basically ready to merge, and my remaining comments are more "nice to have"s.

Have a look and see what you think. I think we can merge this in the next round.

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Update files/en-us/web/css/attr/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

bramus committed:
Move fallback example before color example

bramus committed:
Add compat warnings on top of advanced examples

bramus commented on files/en-us/web/css/attr/index.md:
````diff
@@ -234,6 +274,174 @@ p::after {
+#### HTML
+
+```html
+<div class="cards">
````
.done

bramus committed:
Rework vt example

bramus committed:
Move dimension example before vt example

bramus commented:
Thanks for your extensive input here @chrisdavidmills! I think I have addressed all remarks.

I suggest to update the new “guaranteed-invalid value” glossary page in a standalone PR.

chrisdavidmills committed:
small grammar updates

chrisdavidmills reviewed (approved):
Awesome work @bramus, this is ready to merge!

I just committed a few small grammar nitpicks directly to the branch, rather than bothering you with multiple tiny comments. I hope that's OK.

chrisdavidmills committed:
Merge branch 'main' into advanced-attr

bramus commented:
> I just committed a few small grammar nitpicks directly to the branch, rather than bothering you with multiple tiny comments. I hope that's OK.

Perfect.

Thanks for your thorough review of all this and the successive merge, Chris!
